File: App/Resources/AlgorithmDetails/medianmergesort/medianmergesort.cs
using System;

public class MedianMergeSort
{
  private static void MergeSort(int[] arr, int[] scratch, int start, int end)
  {
    if (end - start < 2) return;
    int middle = (start + end) / 2;
    MergeSort(arr, scratch, start, middle);
    MergeSort(arr, scratch, middle, end);
    int left = start, right = middle, dest = start;
    while (left < middle && right < end)
      scratch[dest++] = arr[left] <= arr[right] ? arr[left++] : arr[right++];
    while (left < middle) scratch[dest++] = arr[left++];
    while (right < end) scratch[dest++] = arr[right++];
    Array.Copy(scratch, start, arr, start, end - start);
  }

  private static int MedianOfThree(int x, int y, int z)
  {
    if (x > y) { int t = x; x = y; y = t; }
    if (y > z) { int t = y; y = z; z = t; }
    if (x > y) { int t = x; x = y; y = t; }
    return y;
  }

  public static void Sort(int[] arr)
  {
    int n = arr.Length;
    int[] scratch = new int[n];
    int start = 0, end = n;
    while (end - start > 16)
    {
      int pivot = MedianOfThree(arr[start], arr[(start + end - 1) / 2], arr[end - 1]);
      int left = start, right = end - 1;
      while (left <= right)
      {
        while (left <= right && arr[left] < pivot) left++;
        while (left <= right && arr[right] > pivot) right--;
        if (left <= right)
        {
          int t = arr[left]; arr[left++] = arr[right]; arr[right--] = t;
        }
      }
      if (left == start || left == end)
      {
        MergeSort(arr, scratch, start, end);
        return;
      }
      if (left - start <= end - left)
      {
        MergeSort(arr, scratch, start, left);
        start = left;
      }
      else
      {
        MergeSort(arr, scratch, left, end);
        end = left;
      }
    }
    for (int i = start + 1; i < end; i++)
    {
      int value = arr[i], j = i;
      while (j > start && arr[j - 1] > value) arr[j] = arr[--j];
      arr[j] = value;
    }
  }

  public static void Main(String[] args)
  {
    int[] array = {
      0, 39, 21, 62, 91, 77, 14, 23,
      90, 69, 51, 81, 68, 83, 32, 56,
      10, 2, 95, 46, 21, 74, 6, 38
    };
    Sort(array);
    string result = "[" + String.Join(", ", array) + "]";
    Console.WriteLine(result);
  }
}

File: App/Resources/AlgorithmDetails/medianmergesort/medianmergesort.go
package main

import "fmt"

func mergeSort(arr, scratch []int, start, end int) {
	if end-start < 2 {
		return
	}
	middle := (start + end) / 2
	mergeSort(arr, scratch, start, middle)
	mergeSort(arr, scratch, middle, end)
	left, right, dest := start, middle, start
	for left < middle && right < end {
		if arr[left] <= arr[right] {
			scratch[dest] = arr[left]
			left++
		} else {
			scratch[dest] = arr[right]
			right++
		}
		dest++
	}
	for left < middle {
		scratch[dest] = arr[left]
		left++
		dest++
	}
	for right < end {
		scratch[dest] = arr[right]
		right++
		dest++
	}
	copy(arr[start:end], scratch[start:end])
}

func sort(arr []int) []int {
	scratch := make([]int, len(arr))
	start, end := 0, len(arr)
	for end-start > 16 {
		x, y, z := arr[start], arr[(start+end-1)/2], arr[end-1]
		if x > y {
			x, y = y, x
		}
		if y > z {
			y, z = z, y
		}
		if x > y {
			y = x
		}
		pivot := y
		left, right := start, end-1
		for left <= right {
			for left <= right && arr[left] < pivot {
				left++
			}
			for left <= right && arr[right] > pivot {
				right--
			}
			if left <= right {
				arr[left], arr[right] = arr[right], arr[left]
				left++
				right--
			}
		}
		if left == start || left == end {
			mergeSort(arr, scratch, start, end)
			return arr
		}
		if left-start <= end-left {
			mergeSort(arr, scratch, start, left)
			start = left
		} else {
			mergeSort(arr, scratch, left, end)
			end = left
		}
	}
	for i := start + 1; i < end; i++ {
		value, j := arr[i], i
		for j > start && arr[j-1] > value {
			arr[j] = arr[j-1]
			j--
		}
		arr[j] = value
	}
	return arr
}

func main() {
	array := []int{0, 39, 21, 62, 91, 77, 14, 23,
		90, 69, 51, 81, 68, 83, 32, 56,
		10, 2, 95, 46, 21, 74, 6, 38}
	fmt.Println(sort(array))
}
